src/VirtoCommerce.MarketingModule.Core/Model/Promotions/DynamicPromotion.cs

```diff
@@ -52,13 +54,17 @@ public override async Task<PromotionReward[]> EvaluatePromotionAsync(IEvaluation
             //Evaluate reward for all promoEntry in context
             foreach (var promoEntry in promoContext.PromoEntries)
             {
-                //Set current context promo entry for evaluation
-                promoContext.PromoEntry = promoEntry;
-
                 foreach (var reward in DynamicExpression?.GetRewards() ?? Array.Empty<PromotionReward>())
                 {
-                    var clonedReward = reward.Clone() as PromotionReward;
-                    EvaluateReward(promoContext, couponIsValid, clonedReward);
+                    PromotionReward clonedReward = reward.Clone() as PromotionReward;
+
```

Comment:
Why not lock just promoContext instead of the surrogate object?

Reply:
Well... Next, I have no idea how to repair the bug with minimal changes, but this solution reduces concurrency. Pay your attention.